source/dem/dem.cc

```diff
@@ -272,24 +284,112 @@ DEMSolver<dim>::cell_weight(
   return 0;
 }
 
+template <int dim>
+unsigned int
+DEMSolver<dim>::cell_weight_with_mobility_status(
+  const typename parallel::distributed::Triangulation<dim>::cell_iterator &cell,
+  const typename parallel::distributed::Triangulation<dim>::CellStatus status,
+  const unsigned int &mobility_status) const
+{
+  // Assign no weight to cells we do not own.
+  if (!cell->is_locally_owned())
+    return 0;
+
+  const unsigned int particle_weight =
+    parameters.model_parameters.load_balance_particle_weight;
+
+  // Applied a factor on the particle weight regards the mobility status
+  // Factor of 1 when mobile cell
+  double alpha = 1.0;
+  if (mobility_status == disable_contact_object.active)
+    {
+      alpha = parameters.model_parameters.active_load_balancing_factor;
+    }
+  else if (mobility_status == disable_contact_object.inactive)
+    {
+      alpha = parameters.model_parameters.inactive_load_balancing_factor;
+    }
+
+  switch (status)
+    {
+      case parallel::distributed::Triangulation<dim>::CELL_PERSIST:
+      case parallel::distributed::Triangulation<dim>::CELL_REFINE:
+        {
+          const unsigned int n_particles_in_cell =
+            particle_handler.n_particles_in_cell(cell);
+          return alpha * n_particles_in_cell * particle_weight;
+          break;
+        }
+
+      case parallel::distributed::Triangulation<dim>::CELL_INVALID:
+        break;
+
+      case parallel::distributed::Triangulation<dim>::CELL_COARSEN:
+        {
+          unsigned int n_particles_in_cell = 0;
+
+          for (unsigned int child_index = 0;
+               child_index < GeometryInfo<dim>::max_children_per_cell;
+               ++child_index)
+            n_particles_in_cell += alpha * particle_handler.n_particles_in_cell(
+                                             cell->child(child_index));
+
+          return n_particles_in_cell * particle_weight;
+        }
+        break;
+
+      default:
+        Assert(false, ExcInternalError());
+        break;
+    }
+
+  return 0;
+}
+
 template <int dim>
 void
 DEMSolver<dim>::setup_background_dofs()
 {
   FE_Q<dim> background_fe(1);
+  background_dh.reinit(triangulation);
   background_dh.distribute_dofs(background_fe);
 }
 
 template <int dim>
 void
 DEMSolver<dim>::load_balance()
 {
+  if (has_disabled_contacts)
+    {
+      std::vector<unsigned int> mobility_status(triangulation.n_active_cells());
```

Comment:
mobility status can be either a map or a vector depending on the section of code you use it. I would name mobility_status_vector and mobility_status_map for the two data types.